Save Logical Type and Coder Registry on cloudpickle save main session

Abacn opened the pull request:
Fix #35738

Restore functionality of user registered logical type and schema, thus preserves the behavior that logical type registry and coder registry are saved to worker when pipeline option `--save_main_session` explicitly added.

**Please** add a meaningful description for your change here

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] Mention the appropriate issue in your description (for example: `addresses #123`), if applicable. This will automatically add a link to the pull request in the issue. If you would like the issue to automatically close on merging the pull request, comment `fixes #<ISSUE NUMBER>` instead.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://github.com/apache/beam/blob/master/CONTRIBUTING.md#make-the-reviewers-job-easier).

To check the build health, please visit [https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md](https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md)

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[![Build python source distribution and wheels](https://github.com/apache/beam/actions/workflows/build_wheels.yml/badge.svg?event=schedule&&?branch=master)](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[![Python tests](https://github.com/apache/beam/actions/workflows/python_tests.yml/badge.svg?event=schedule&&?branch=master)](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[![Java tests](https://github.com/apache/beam/actions/workflows/java_tests.yml/badge.svg?event=schedule&&?branch=master)](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)
[![Go tests](https://github.com/apache/beam/actions/workflows/go_tests.yml/badge.svg?event=schedule&&?branch=master)](https://github.com/apache/beam/actions?query=workflow%3A%22Go+tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI or the [workflows README](https://github.com/apache/beam/blob/master/.github/workflows/README.md) to see a list of phrases to trigger workflows.

Abacn commented:
Tested with Dataflow, registered a user logical type and a namedtuple coder in main session, and checked logical type/coder registry had these items in DoFn process (job id): 2025-09-25_11_50_50-8758466825726433704

shunping commented on sdks/python/apache_beam/internal/cloudpickle_pickler.py:
```diff
@@ -196,12 +196,35 @@ def _lock_reducer(obj):
+    pickler = cloudpickle.CloudPickler(file)
+    # TODO(https://github.com/apache/beam/issues/18500) add file system registry
+    # once implemented
+    pickler.dump({"coder": coder_reg, "logicaltype": logicaltype_reg})
```
"logical_type"

Abacn commented:
> we can probably revert some JDBC logical type fixes we had before

These fixes stay there as previously it relied on io.jdbc gets imported to make those type work, which was error prone

> Btw, is this enabled by default for cloudpickle or do users need to set the `--save_main_sesson` flag?

Not by default. The goal of this PR is to restore the save_main_sesson behavior pre 2.65.0 (make it effective again). Enable "save_main_sesson" by default for cloudpickle is left for future discussion as it affects users using default pipeline options.

Abacn commented on sdks/python/apache_beam/internal/cloudpickle_pickler.py:
```diff
@@ -196,12 +196,35 @@ def _lock_reducer(obj):
+
+  with _pickle_lock, open(file_path, 'wb') as file:
+    coder_reg = typecoders.registry.get_custom_type_coder_tuples()
+    logicaltype_reg = schemas.LogicalType._known_logical_types.copy()
```
thanks, changed

Abacn commented on CHANGES.md:
```diff
@@ -94,6 +94,10 @@
 * PulsarIO has now changed support status from incomplete to experimental. Both read and writes should now minimally
   function (un-partitioned topics, without schema support, timestamp ordered messages for read) (Java)
   ([#36141](https://github.com/apache/beam/issues/36141)).
+* (Python) Logical type and coder registry are saved for pipelines with `save_main_session` pipeline option enabled in
```
I agree it becomes misleading naming, unfortunately this flag is named by implementation detail instead of what user want to achieve (states at submission session env are preserved job execution time). Here I want to un-break existing pipeline option setting. If user needs a new flag on upgrading their pipeline they can choose to add `--pickle_library=dill`. Probably one can add an alias "--save_beam_registries" and recommend for use for new users.

Making it enabled all times is a step further and has risk of new breakages, though I haven't observe

claudevdm replied:
https://github.com/apache/beam/blob/bb340c2f66ac8730334160d6ed5ecd18822d059d/sdks/python/apache_beam/options/pipeline_options.py#L1647

I am still worried that changing the default behavior for cloudpickle when `save_main_session=True` which used to be a no-op (since cloudpickle doesnt need main session to unpickle types defined in main) can lead to some kind of errors.

I think having an explicit flag so that users who run into a logical type issue can use is reasonable instead of pickling registries for all users that unintentionally set the `save_main_session=True` flag but don't use custom logical types at all.

claudevdm replied:
If we want to unbreak existing pipelines that use custom logical types, can we detect that someone registered a logical type (from outside schemas.py) and set the `save_beam_registries=True`?

claudevdm commented on sdks/python/apache_beam/internal/cloudpickle_pickler.py:
```diff
@@ -196,12 +196,35 @@ def _lock_reducer(obj):
+  # Since References are saved (https://s.apache.org/beam-picklers), we only
+  # dump supported Beam Registries (currently only logical type registry)
+  from apache_beam.typehints import schemas
+  from apache_beam.coders import typecoders
```
Does it matter if the coder is registered on the worker though? My understanding is `coders.registry.register_coder(JdbcTestRow, coders.RowCoder)` just maps any transforms that use JdbcTestRow to use RowCoder during pipeline construction.

Then the RowCoder is pickled as part of the pipeline proto, and the worker doesnt actually need the coder registry, it just uses the pickled coder.

Am I missing something? Can you reproduce some sort of error in the case where coder registry is not correct on worker?

Abacn replied:
That's true. It doesn't have error, meaning "worker doesnt actually need the coder registry" in this use case. The original error is due to logical type registry, adding coder registry here is for completeness coneptually as both are exposed to user.

claudevdm commented on sdks/python/apache_beam/typehints/schemas.py:
```diff
@@ -704,8 +709,24 @@ def copy(self):
+      copy.by_urn[urn] = logical_type
+      copy.by_logical_type[logical_type] = urn
+      copy.by_language_type[logical_type.language_type()] = logical_type
+      copy._custom_urns.add(urn)
```
return copy?

claudevdm reviewed (changes requested):
Thanks!

1. Can we add an integration test/validatesrunner test?
2. Can we dump/load session by default if we detect that the user has used custom logical types? I dont think it is necessary for a user to control this via flag if we can detect, but maybe we can keep an option to explicitly disable it in case there are issues?

Abacn commented:
> Can we add an integration test/validatesrunner test?

Added a unit test. Tried to add a local integration test (run on prism runner) but find that Prism does not save/load sessions. Actually only service_runner (DataflowRunner) do.

Now prefer to revert #35659 (and other workarounds) in follow ups, then the path is exercised by integration tests previously affected

> Can we dump/load session by default if we detect that the user has used custom logical types? I dont think it is necessary for a user to control this via flag if we can detect, but maybe we can keep an option to explicitly disable it in case there are issues?

Done. load session is now enabled by default for cloudpickle and introduced a "disable_save_main_session" flag to explicitly disable it
